src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -544,24 +554,123 @@ void MarkCustomAttributes (ICustomAttributeProvider provider, in DependencyInfo
 				MarkCustomAttribute (ca, reason);
 				MarkSpecialCustomAttributeDependencies (ca, provider);
 			}
+
+			if (!(provider is MethodDefinition || provider is FieldDefinition))
+				return;
+
+			var dynamicDependencies = _context.Annotations.GetLinkerAttributes<DynamicDependency> ((IMemberDefinition) provider);
+			Debug.Assert (dynamicDependencies != null);
+
+			foreach (var dynamicDependency in dynamicDependencies)
+				MarkDynamicDependency (dynamicDependency, (MemberReference) provider);
 		}
 
 		protected virtual bool ProcessLinkerSpecialAttribute (CustomAttribute ca, ICustomAttributeProvider provider, in DependencyInfo reason)
 		{
-			if (IsUserDependencyMarker (ca.AttributeType) && provider is MemberReference mr) {
+			var isPreserveDependency = IsUserDependencyMarker (ca.AttributeType);
+			var isDynamicDependency = LinkerAttributesInformation.IsAttribute<DynamicDependencyAttribute> (ca.AttributeType);
+
+			if (!((isPreserveDependency || isDynamicDependency) && provider is MemberReference mr))
+				return false;
+
+#if !FEATURE_ILLINK
+			if (isPreserveDependency)
 				MarkUserDependency (mr, ca);
+#endif
 
-				if (_context.KeepDependencyAttributes || Annotations.GetAction (mr.Module.Assembly) != AssemblyAction.Link) {
-					MarkCustomAttribute (ca, reason);
-				} else {
-					// Record the custom attribute so that it has a reason, without actually marking it.
-					Tracer.AddDirectDependency (ca, reason, marked: false);
+			if (_context.KeepDependencyAttributes || Annotations.GetAction (mr.Module.Assembly) != AssemblyAction.Link) {
+				MarkCustomAttribute (ca, reason);
+			} else {
+				// Record the custom attribute so that it has a reason, without actually marking it.
+				Tracer.AddDirectDependency (ca, reason, marked: false);
+			}
+
+			return true;
+		}
+
+		void MarkDynamicDependency (DynamicDependency dynamicDependency, MemberReference context)
+		{
+			Debug.Assert (context is MethodDefinition || context is FieldDefinition);
+			AssemblyDefinition assembly;
+			if (dynamicDependency.AssemblyName is string assemblyName) {
+				assembly = _context.GetLoadedAssembly (assemblyName);
+				if (assembly == null) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Unresolved assembly '{assemblyName}' in DynamicDependencyAttribute on '{context}'", 2031));
```

Comment:
Will that ever happen when you do same check during load?

Reply:
Unfortunately it could happen if an assembly added by `DynamicDependencyLookupStep` has further `DynamicDependencyAttributes` that depend on assemblies not in the context, because we don't do the lookup recursively.